Add grpc client and server plugin

nunofgs opened the pull request:
Hi DataDog team!

This PR adds automatic instrumentation support for grpc, both client and server. The `client` plugin is implemented using client interceptors and should be pretty resilient to upstream changes.

The `server` plugin however, has no such support and I took a stab at implementing — finding any injection point as best I could. It is fully functional but I cannot guarantee that it'll be as resilient as grpc-node evolves.

<img width="1008" alt="dd-trace grpc" src="https://user-images.githubusercontent.com/92085/56892850-c5e13000-6a78-11e9-8cac-6072434f7f89.png">

Let me know what you think!

* Fixes #311 

Note: `require-dir` was removed since it doesn't support nested folders. There was already a file (`src/plugins/index.js`) with all of the supported plugins which I believe is even more appropriate to use.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+                const values = config.arguments(status.metadata.getMap())
+
+                for (const key in values) {
+                  span.setTag(`error.metadata.${key}`, values[key])
```
Why only include these when there is an error?

nunofgs replied:
Hehe, nicely caught. That's because I've personally never sent metadata on an `OK` response (only for errors) but it makes sense to allow for all responses. Updated.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+              } else {
+                span.setTag(`error.metadata`, status.metadata.getMap())
+              }
+            }
```
This block should be refactored as a function as it's used in multiple places.

nunofgs replied:
There's two instances in this file but always with different prefixes. Do you mean create a utility function and refactor all plugins (ex: [graphql](https://github.com/DataDog/dd-trace-js/blob/6b26a65268a1b726b5716ee6cabff5ae162409be/src/plugins/graphql.js#L258-L261))?

rochdev replied:
I just meant in general to use more functions when possible to reuse code that is similar or identical and to make things cleaner. Feel free to ignore however, as I already have a task to clean up integrations in general to make it easier to contribute.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+  name: 'grpc',
+  versions: ['>=1.13'],
+  patch,
+  unpatch
```
It would be easier to read to clearly separate the patch from the wrapper factory, the wrapper and the wrapped function, similar to https://github.com/DataDog/dd-trace-js/blob/master/src/plugins/cassandra-driver.js

I know we have some integrations that don't do this 😅

rochdev replied:
Similar to the above, feel free to ignore for now.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+
+module.exports = [{
+  name: 'grpc',
+  versions: ['>=1.13'],
```
This version is somewhat new. Any reason why 1.x cannot be supported?

nunofgs replied:
1.13 was when grpc [deprecated](https://github.com/grpc/grpc-node/releases/tag/grpc%401.13.0) `grpc.load()` in favor of `@grpc/proto-loader`.

I could conceivably lower to 1.11 (lowest I could possibly go since it was when client_interceptors were added) but we'd get warnings emitted when running tests of the latest version, prompting us to use `@grpc/proto-loader`.

I might be able to support 1.11 by using `grpc.load()` or `@grpc/proto-loader` depending on the version being tested. Want me to give it a shot?

rochdev replied:
It sounds like possibly a lot of work to gain only 2 minor versions. I would say to keep it at 1.13 for now and we can revisit after it's released.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  this.wrap(grpc.Client.prototype, 'resolveCallInterceptors', resolveCallInterceptors => {
+    return function () {
```
Preserve the original name suffixed with `withTrace` for stack traces.

nunofgs replied:
Good call :+1:

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,137 @@
+
+function createWrapHandler (grpc, tracer, config, handler) {
+  return function wrapHandler (func) {
+    return function funcHandler (emitter) {
```
Preserve the original function name with `withTrace` suffix for stack traces.

nunofgs replied:
Gotcha :+1:

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,137 @@
+
+function handleError (config, span, err, metadata) {
+  span.addTags({
+    'error.code': err.code,
```
Any way to also get the type? If not, hardcode `error.type` to `Error`.

nunofgs replied:
Added 👍

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,137 @@
+      if (request) {
+        if (metadata) {
+          for (const [key, value] of Object.entries(metadata.getMap())) {
+            span.setTag(`grpc.metadata.${key}`, value)
```
Can metadata contain sensitive data?

nunofgs replied:
Potentially, yes. It's user input as far this lib is concerned. I see what you're saying.

I'll add another callback to the config option.

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,137 @@
+        // Finish the span if the response only if it was successful.
+        // Otherwise it'll be finished in the `error` listener.
+        emitter.on('finish', function () {
+          if (emitter.status.code === 0) {
```
Is it possible that the status is not `0` and the `error` event is not emitted?

rochdev replied:
Also, what happens if the request is cancelled?

nunofgs replied:
> Is it possible that the status is not 0 and the error event is not emitted?

Hard to say but from looking at the [server implementation](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/server.js) I'd consider it a bug if that's possible.

> Also, what happens if the request is cancelled?

Unfortunately I can't seem to find a way to support cancelled requests. A cancelled unary call will [never call](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/server.js#L587-L589) our wrapped function.

It would be great if we could patch the [call.startBatch()](https://github.com/grpc/grpc-node/blob/b1e7d446265065f35471dc533e25db5a8ab4f5d9/packages/grpc-native-core/src/server.js#L575) or the [handleUnaryCall()](https://github.com/grpc/grpc-node/blob/b1e7d446265065f35471dc533e25db5a8ab4f5d9/packages/grpc-native-core/src/server.js#L567) calls but I just couldn't find an injection point to patch them. Totally open to suggestions if you have an idea on this one.

rochdev replied:
This is unfortunately a blocker, since not finishing a span will result in a memory leak. I would have to take a look, but hopefully there is a construct available to detect that.

Is the emitter also an actual stream? If that's the case maybe it emits some of the standard events on cancellation?

rochdev replied:
A test for this is definitely needed as well.

nunofgs replied:
Makes sense. I'll see if I can come up with a way to handle cancellations. Will get back to you soon.

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,137 @@
+        }
+      }
+
+      if (isStream) {
```
Does this handle all the different method types?

nunofgs replied:
I believe so. There's only 4 in total (bidi, client_stream, server_stream and unary), two of which are streams and covered by this condition.

rochdev replied:
I'm not 100% sure, but I believe `bidi` might be a stream as well?

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,137 @@
+      for (const handler in this.handlers) {
+        self.wrap(this.handlers[handler], 'func', createWrapHandler(grpc, tracer, config, handler))
+
+        patchedHandlers.push([this.handlers[handler], 'func'])
```
Using an object would make this clearer.

rochdev replied:
Also, why is the name needed if it's always `func`?

nunofgs replied:
> Using an object would make this clearer.

I gave it a try but I find it a bit more unclear:

```js
patchedHandlers[handler] = this.handlers[handler]
```

Also, it forces me to iterate over the object's values in the `unpatch` method and completely discard the object's keys. (which is why I went for an array 😇 )

Do you have a suggestion here?

> Also, why is the name needed if it's always func?

I was initially patching multiple functions and needed a generic solution but have since managed to find the `func` injection point. I'll hard-code in the unpatch method. 👍

rochdev replied:
> Also, it forces me to iterate over the object's values in the unpatch method and completely discard the object's keys.

I meant something like `patchedHandlers.push({ handler, name })`. Then when you iterate it's easier to understand as you are accessing a property with proper semantics instead of an index.

This may now be a non-issue though if you are going to remove the name since it's no longer necessary.

rochdev commented:
Thanks for the PR! I did a first pass of review but in general the instrumentation itself looks good.

I scheduled this for 0.12.0 since we don't plan to release it before at least a couple weeks, so we should be able to include this feature in the release.

rochdev commented:
> It is fully functional but I cannot guarantee that it'll be as resilient as grpc-node evolves.

We have tests that run nightly so we should be able to find anything that breaks pretty quickly.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+          const values = config.arguments(message)
+
+          for (const key in values) {
+            span.setTag(`grpc.arguments.${key}`, values[key])
```
Looks like these are called `fields` according to the [gRPC guide](https://grpc.io/docs/guides/)

rochdev replied:
Since these are part of the message, I would probably go with `grpc.message.fields.*`

nunofgs replied:
Ah, I struggled a bit with that as well but I found that `message` was a bit ambiguous as to whether it represents the request or the response message (in the case of the `client`, I'm only setting tags for the request message)

I ended up going with `arguments` since it's what `grpc` is calling internally to their variable. See [here](https://github.com/grpc/grpc-node/blob/38637b51038e3c0a7dae8339c7d5fbeab888adb4/packages/grpc-native-core/src/client.js#L30), [here](https://github.com/grpc/grpc-node/blob/309033f681d535c3ac31ce85ae6734f6bec7b470/packages/grpc-js/src/client.ts#L164) and [here](https://github.com/grpc/grpc-node/blob/309033f681d535c3ac31ce85ae6734f6bec7b470/packages/grpc-js/src/client.ts#L137).

I'm fine with changing to `grpc.message.fields.*` but I'd also have to change the callback in the config. What name would you suggest in that case? (`config.fields` for example?)

nunofgs commented:
Thanks for the in-depth review @rochdev and willingness to accept the PR :) !

I'll update the branch with your comments addressed shortly.

nunofgs commented on scripts/install_plugin_modules.js:
```diff
@@ -68,6 +67,10 @@ function assertFolder (name, version) {
     fs.mkdirSync(folder())
   }
 
+  if (name && name.includes(path.sep)) {
```
This is meant to support the `@grpc/proto-loader` dependency defined in the externals.json file. Without it you get an error since we'll be trying to "mkdir" a folder with a path separator.

rochdev replied:
Thanks for clarifying, I forgot about the externals.

nunofgs commented on scripts/install_plugin_modules.js:
```diff
@@ -2,11 +2,10 @@
 const semver = require('semver')
 const exec = require('./helpers/exec')
-const plugins = requireDir('../src/plugins')
+const plugins = require('../src/plugins')
```
Good point! Done.

nunofgs commented on test/plugins/agent.js:
```diff
@@ -141,7 +141,7 @@ module.exports = {
   wipe () {
     const basedir = path.join(__dirname, '..', '..', 'versions')
-    const exceptions = ['/libpq/'] // wiping native modules results in errors
+    const exceptions = ['/libpq/', '/grpc/'] // wiping native modules results in errors
```
Unfortunately it won't work. There is no `grpc-native-core` dependency in this case. The folder you're [most likely thinking of](https://github.com/grpc/grpc-node/tree/master/packages/grpc-native-core) is actually the folder that gets published to npm as the `grpc` package.

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+
+  config = config.client || config
+
+  function gRPCInterceptor (options, nextCall) {
```
Good call. Updated.

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+      childOf: scope && scope.span(),
+      tags: {
+        [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_CLIENT,
+        'grpc.method_name': path,
```
Most other plugins only go 2 levels deep so I was trying to keep that consistency, but I agree that it'll be cleaner. Updated.

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+      tags: {
+        [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_CLIENT,
+        'grpc.method_name': path,
+        'grpc.method_type': responseStream ? 'server_stream' : 'unary',
```
Missed those! Thanks!

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+        'grpc.method_name': path,
+        'grpc.method_type': responseStream ? 'server_stream' : 'unary',
+        'resource.name': path.split('/').slice(-1),
+        'service.name': config.service || `${tracer._service}-grpc`
```
:+1:

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+        [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_CLIENT,
+        'grpc.method_name': path,
+        'grpc.method_type': responseStream ? 'server_stream' : 'unary',
+        'resource.name': path.split('/').slice(-1),
```
Personal preference but totally willing to change if you prefer it. My thinking was:

* `grpc.method.name` would be something like: `/FooBarService/GetFooBar`
* `resource.name` would be: `GetFooBar`

If we keep the entire path we might as well remove the `grpc.method.name` tag since it would be duplicated.

See screenshot below to see how it looks (in this case I have a "GetUser" resource name):

<img width="1008" alt="dd-trace grpc" src="https://user-images.githubusercontent.com/92085/57109667-fe5d6400-6d2d-11e9-8d47-d3cba9782747.png">

rochdev replied:
It's fine to keep the tag as well. It's possible that we change the value of `resource.name` eventually for example but the method name will always be the same.

My main concern with removing the service name from the resource is that it could cause stats to not be calculated correctly if multiple services have the same method. Right now it wouldn't cause issues but we are planning to add more calculations soon that could be impacted.

nunofgs replied:
> My main concern with removing the service name from the resource is that it could cause stats to not be calculated correctly if multiple services have the same method

Would that not apply to the graphql and koa plugins as seen on the screenshot? (`me` and `POST`, respectively)

rochdev replied:
For the server possibly since then you know for sure what the service is. But for a remote endpoint you don't have this information.

rochdev replied:
@nunofgs Do you think the proposed change makes sense?

nunofgs replied:
Whoa, totally missed your comment. Just to clarify, you're requesting:

```js
        'grpc.method_name': path,
        'resource.name': path
```

Correct?

rochdev replied:
Correct.

nunofgs replied:
PR Updated but fwiw I still don't see much value in sending both tags with the same content. Will defer to your judgement here :)

rochdev replied:
`resource.name` is not really a tag once it's been formatted by the tracer and becomes a field that has a special meaning in our backend and UI. This means it could be altered later in the pipeline for example. It also means its value will be removed from the list of tags. Having a separate tag with the same value ensures the value is still available as an untouched tag.

It's also possible that we eventually stop relying on the `resource` field and switch to use the tag directly instead. In that case, having the tag already available will reduce the overhead of the migration.

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+            span.setTag(`grpc.arguments.${key}`, values[key])
+          }
+        } else {
+          span.setTag(`grpc.arguments`, message)
```
👍

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+
+    return new grpc.InterceptingCall(nextCall(options), {
+      sendMessage: (message, next) => {
+        if (config.arguments) {
```
Sorry, not sure if I understood what you mean. I was basing this functionality on the graphql plugin's `variables` [callback](https://github.com/DataDog/dd-trace-js/blob/6b26a65268a1b726b5716ee6cabff5ae162409be/src/plugins/graphql.js#L257-L262). 

Can you elaborate?

rochdev replied:
Even though it's not apparent at a glance, the `variables` config from GraphQL accepts more than just a callback: https://github.com/DataDog/dd-trace-js/blob/6b26a65268a1b726b5716ee6cabff5ae162409be/src/plugins/graphql.js#L405-L414

This allows to use a simple array of strings for simple cases.

nunofgs replied:
Ah, gotcha. Will do.

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+        const tracingMetadata = metadata || new grpc.Metadata()
+        const meta = {}
+
+        tracer.inject(span, FORMAT_HTTP_HEADERS, meta)
```
The `tracingMetadata` here is an instance of the `grpc.Metadata` class. It doesn't support setting a property (you need to call `.set()` as below)

rochdev replied:
Makes sense. I would probably extract this part to a function to make this cleaner. Also, since gRPC is not HTTP, `FORMAT_TEXT_MAP` would be more correct semantically, even do it's equivalent to `FORMAT_HTTP_HEADERS`.

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,100 @@
+            if (status.code !== 0) {
+              span.addTags({
+                'error.code': status.code,
+                'error.msg': status.details
```
> This will conflict with the predefined error object that we usually send. Any reason not to use status.code and status.details directly, and simply set error to true?

🤔 I was attempting to provide the functionality of displaying the correct error message in Datadog by setting `error.code` and `error.msg` here. Is that not how it works? (ex: from the [graphql plugin](https://github.com/DataDog/dd-trace-js/blob/master/src/plugins/graphql.js#L316-L322))

> These could probably always be included as well if they don't contain the word "error".

I could include them but `status.code` is the gRPC code which will always be `0` for a successful message (see [here](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/constants.js#L32) and [here](https://github.com/grpc/grpc-node/blob/633736225ac4f8ce05bd246ffe576616e9290615/packages/grpc-native-core/src/client_interceptors.js#L84)) so I didn't really find it interesting to include. Happy to add it if you feel it's useful.

Also, `status.details` is only set if there's a non-zero status code.

rochdev replied:
I agree that `status.code` and `status.details` are not necessary in that context. We can always add them later if necessary.

About the error tags, basically our backend extracts `error.type`, `error.type` and `error.stack` as metadata and removes the tags. This means that the result of the above will be an `error.code` tag and an error message in the UI. I'm not 100% sure if `error.type` is required, but we can always add it later if the error doesn't appear as expected.

rochdev replied:
Actually I would probably go with `status.code` + `error.msg` in that case, otherwise it seems like there was some kind of error object. `status.code` seems to be more semantically correct. It's actually a status code, the fact that it represents an error is irrelevant for that tag specifically. Basically the same logic as `http.status_code`.

nunofgs replied:
Makes sense. I'll update.

nunofgs commented:
> Just noticed that the tests are not running in CI and should be added to .circleci/config.yml

Missed that, thank you. Added.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,117 @@
+
+    return new grpc.InterceptingCall(nextCall(options), {
+      sendMessage: (message, next) => {
+        if (config.message_fields) {
```
Our standard for keys is camelCase. However, I think just `fields` would be enough since fields only exist on messages.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,117 @@
+    const path = options.method_definition.path
+    const methodType = getMethodType(options.method_definition)
+    const scope = tracer.scopeManager().active()
+    const span = tracer.startSpan('grpc.client', {
```
This should be `grpc.request` to be consistent with other integrations we well as gRPC's own terminology.

rochdev commented:
Ignore the `node-mongodb-core` errors, it's just noise because of an issue we are having with CircleCI. Postgres will need to be fixed however, but I'm not sure why it's failing. It must be related to adding `grpc` to the list of native modules. If you can't figure it out let me know and I will take a look when the PR is otherwise ready to merge.

nunofgs commented:
> Postgres will need to be fixed however, but I'm not sure why it's failing. It must be related to adding grpc to the list of native modules. If you can't figure it out let me know and I will take a look when the PR is otherwise ready to merge.

Found it. `prev` wasn't being used [here](https://github.com/DataDog/dd-trace-js/blob/ba9be7c92cdb25a373f4a81425dff0565fcae89e/test/plugins/agent.js#L148). I ended up changing the implementation a bit but I can go back to an Array.reduce() if you prefer.

nunofgs commented:
@rochdev sorry for taking so long to get back to this. Had a busy week.

I've updated the PR with support for cancellation and bidi streams (as well as tests of course). I believe all actionable comments should be addressed.

Mind doing another review?

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,143 @@
+      },
+
+      cancel: () => {
+        span.setTag('cancelled', true)
```
This should be scoped like the other tags. Also, would it be better to have some kind of "state" with different values instead of multiple tags per state?

nunofgs replied:
I managed to get a nice textual status of each request for almost every case, except _unary_ calls on the server plugin. :(

Spent a lot longer on it than I care to admit 🙈but can't really find a solution the way that grpc is set up. Hopefully when we have server interceptors it'll improve.

rochdev replied:
A few things about the new implementation:

• Would it be possible to use the keys of the constants instead of strings?
• Can something like [this](https://github.com/grpc/grpc-node/blob/633736225ac4f8ce05bd246ffe576616e9290615/packages/grpc-native-core/src/server.js#L44-L70) be used to categorize errors?
• Since the code can be set by the user to a value outside of the known codes, I would say it would be best to have the code as a number instead of a string. We could always include the string as well for known codes in another tag.

nunofgs replied:
> Can something like this be used to categorize errors?

Sorry, not sure what you mean here. 

> Since the code can be set by the user to a value outside of the known codes, I would say it would be best to have the code as a number instead of a string.

Makes sense. I've updated the PR.

> We could always include the string as well for known codes in another tag.

Agreed. I've left it out for now but it should be trivial to add if needed.

rochdev replied:
> Sorry, not sure what you mean here.

I think I misunderstood what you meant by `I managed to get a nice textual status of each request for almost every case, except unary calls on the server plugin.`

Does this mean that for unary calls it's not possible to get the status code at all? How useful would the status code be for these calls?

nunofgs replied:
> Does this mean that for unary calls it's not possible to get the status code at all?

It's possible on the _client_ plugin (since the client interceptor API exposes the status code) but not on the _server_ plugin.

> How useful would the status code be for these calls?

We know whether the unary call returned an error, so:

1. If it didn't return an error (was successful) we can pretty much guarantee it was OK (code `0`).
2. If it **did** return an error, I was defaulting to `UNKNOWN` but it's not technically correct. It could also be INVALID_ARGUMENT, NOT_FOUND, etc.

rochdev replied:
My understanding is that the status code comes from the error object for errors, so it may be available at that point. Not 100% sure though.

nunofgs replied:
That's technically true but the function I'm wrapping doesn't give me a chance to grab the error object with the code. I'd need to be able to grab the `err` variable from [here](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/server.js#L575-L590), but I'm wrapping the `func()` call which doesn't get access to it. 

At the point where my function is called, `emitter.error` doesn't have a code property.

rochdev replied:
Would wrapping `call.startBatch` as well work in this case?

rochdev replied:
Looks like it would be difficult to do that properly. In that case I'd probably still check for `error.code` since it could have been set elsewhere, and if there is no `error.code` then skip setting the tag and rely only on the information provided by the error tags.

nunofgs replied:
I spent a lot of time trying to wrap `call.startBatch` early on (as well as the `handleUnary`, `handleServerStreaming`, etc calls but with little success :(

While grpc is indeed attempting to copy the `code` from the error object ([if it exists](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/server.js#L53)) I cannot reproduce this behaviour in a test.

From what I can gather it's meant to be provided by the caller so I guess it makes sense to honour it if it's there.

PR updated.

rochdev replied:
I prefer to not set the tag instead of using `UNKNOWN` when `error.code` is not set since it could be confusing if there is actually a status code but it's different.

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,161 @@
+          span.finish()
+        })
+
+        // Finish the span if the response only if it was successful.
```
Nit: `span of the response`

rochdev commented:
@nunofgs One thing I'm wondering is about the context propagation. We are currently working on an HTTP2 integration which will handle context propagation, but right now you are doing it directly in gRPC. What do you think should be the expected behavior? Should we just drop propagation from gRPC at that point or is there still value to have it there?

nunofgs commented:
> @nunofgs One thing I'm wondering is about the context propagation. We are currently working on an HTTP2 integration which will handle context propagation, but right now you are doing it directly in gRPC. What do you think should be the expected behavior? Should we just drop propagation from gRPC at that point or is there still value to have it there?

🤔 gRPC uses http2 as transport protocol but it supports others and it's also possible to write your own. It might be useful to keep the gRPC context propagation for those cases.

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,171 @@
+          span.finish()
+        })
+
+        // Finish the span if the response only if it was successful.
```
```suggestion
        // Finish the span of the response only if it was successful.
```

rochdev commented on src/plugins/grpc/server.js:
```diff
@@ -0,0 +1,166 @@
+
+        // Unfortunately it doesn't seem possible to find the actual status code of the request at this point,
+        // so we're using these defaults until grpc adds proper support for server interceptors.
+        span.setTag('grpc.request.status', err ? grpc.status.UNKNOWN : grpc.status.OK)
```
It's possible the error object contains a `code` property as well. It should be used when that's the case.

Reference: https://github.com/grpc/grpc-node/blob/633736225ac4f8ce05bd246ffe576616e9290615/packages/grpc-native-core/src/server.js#L53-L58

nunofgs replied:
Unfortunately `handleError` is only called later in the lifecycle. Essentially, it's called within the callback that I execute, [here](https://github.com/grpc/grpc-node/blob/633736225ac4f8ce05bd246ffe576616e9290615/packages/grpc-native-core/src/server.js#L595).

rochdev replied:
Yes but if `handleError` sets `status.code` to the value of `error.code`, and `error.code` was already set at the point `handleError` is called it's possible that it's already set when this code runs as well.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,144 @@
+      },
+
+      cancel: () => {
+        span.setTag('grpc.request.status', grpc.status.CANCELLED)
```
The tag should probably be `grpc.request.status.code` to allow future extensions to `grpc.request.status`.

rochdev replied:
Just realized that it's also `grpc.request.status.code` for the server. Is that the expected name? Shouldn't it be `response`?

I wonder if just `grpc.status.code` would be simpler since there is a single status code for the entire request on response. This is also what we do for HTTP.

nunofgs commented:
@rochdev I pushed the latest changes a few days ago. Let's do a final pass? :)

rochdev commented:
@nunofgs Will do a last pass today. Expect a few nitpicks since this will be the last pass before I merge this :)

rochdev commented:
One thing I forgot in the review: Right now the scope is activated in the server but not the client. This means that any child spans like `http2.request`, `tcp.connect` and `dns.lookup` will end up as children of the parent.

rochdev commented:
@nunofgs Just a heads up that we're planning a release in the next few days. We can include this feature in the release but the requested changes would have to be completed by then so we can merge.

nunofgs commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,144 @@
+      childOf: scope && scope.span(),
+      tags: {
+        [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_CLIENT,
+        'grpc.method.name': path,
```
That's very similar to my [initial approach](https://github.com/DataDog/dd-trace-js/pull/547#discussion_r280627448).

Note that grabbing the "first part of the path" as suggested may have its own issues. In the example image I pasted above, the gRPC service is not namespaced to a package. In that case we should also be parsing the `namespace` out of the path.

I can try to figure out all of the possible permutations and parse the `path` out to multiple tags if you like. What do you think?

rochdev replied:
The alternative would be to simply not include the tags at all for now, but if it's not too difficult it would definitely be best to have them now.

rochdev commented on src/plugins/grpc/client.js:
```diff
@@ -0,0 +1,143 @@
+          const values = configFields(message)
+
+          for (const key in values) {
+            span.setTag(`grpc.request.message.fields.${key}`, values[key])
```
What happens if there are multiple messages in the request? It looks like this should be attached to the message span but it doesn't exist in this implementation.

@nunofgs Is this something you need? If not I would say to just leave it out for this initial implementation. Otherwise a new message span will probably need to be introduced to hold these tags.

The same applies to the server as well.

nunofgs replied:
Not sure I know what you mean. A span is being created upon each grpc message, both on the client and server. Maybe you mean something else?

rochdev replied:
My understanding from other languages is that there is a single request (or stream) for multiple messages. A span is created on the stream and may contain multiple messages, which may or may not have spans, but since there is a single span created here I assumed it was for the stream and not the message, especially since it handles propagation. Is this not correct?

nunofgs replied:
🤔 from what I can see, client interceptors are called on every message. Even if there are multiple messages per request, a client interceptor call is scoped to a single method which is where the span is being created, right?

rochdev replied:
I think the safest would be to have a test for streams with multiple messages. This will make it clear what the behavior is and will ensure that the implementation hooks in the right place when there are multiple messages.

In other languages we use different spans for the request and the messages, so this is important to ensure interoperability.

rochdev commented:
@nunofgs We've merged HTTP2 and would like to release gRPC at the same time. Let me know if you need any help bringing this to the finish line. I left a few small comments above but otherwise we should be able to release this within days.

I'll handle the merge since we changed the project structure.

Thanks a lot for this PR and sorry for the swarm of comments 😅

nunofgs commented:
> @nunofgs We've merged HTTP2 and would like to release gRPC at the same time. Let me know if you need any help bringing this to the finish line. I left a few small comments above but otherwise we should be able to release this within days.

Sounds good. I think I've addressed all comments but let me know if I missed anything. 

> I'll handle the merge since we changed the project structure.

Quite a big change. I noticed. I'm happy to rebase this PR if you want!

> Thanks a lot for this PR and sorry for the swarm of comments 😅

No problem! It's important we get this plugin right. I imagine a lot of people will use it :)

rochdev commented:
@nunofgs We would like to release this soon. Is it ok if we take over the PR? It will mostly be a few minor changes, merging with the new structure, and removing message metadata. We can revisit message support in a later iteration.

nunofgs commented:
@rochdev not at all. Please go ahead. I haven't had time to pick this back up but I agree that message support can come at a later date.

rishabh commented on packages/datadog-plugin-grpc/src/server.js:
```diff
@@ -0,0 +1,136 @@
+const kinds = require('./kinds')
+const { addMethodTags, addMetadataTags, getFilter } = require('./util')
+
+function handleError (span, err) {
```
Are the `err`'s a typeof `Error`? If so, we can just let the formatter handle the error tags.

rishabh commented on packages/datadog-plugin-grpc/src/client.js:
```diff
@@ -0,0 +1,173 @@
+  return span
+}
+
+function ensureMetadata (args, grpc) {
```
We can simplify this by using a single `splice`, if we don't care about the original array, or two `slice`'s if we care.

rochdev replied:
Using `Array` methods on an `arguments` object may cause certain v8 optimizations to be disabled.

rishabh replied:
Ah, ok, sounds good!

rochdev commented:
### Plugin Checklist
<!-- Fill this section if adding or updating a plugin. Remove otherwise. -->

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

rochdev commented on scripts/install_plugin_modules.js:
```diff
@@ -22,7 +22,6 @@ function run () {
 function assertVersions () {
   let filter = []
   let names = Object.keys(plugins)
-    .filter(key => key !== 'index')
```
It's no longer useful now that `require-dir` is not used, meaning there is no `index` key that needs to be filtered out anymore.

rishabh replied:
Dope!

rishabh commented:
Let's add some tests to ensure that we aren't modifying any metadata (except for the context propagation).

rochdev commented on packages/datadog-plugin-grpc/src/util.js:
```diff
@@ -0,0 +1,51 @@
+const log = require('../../dd-trace/src/log')
+
+module.exports = {
+  addMethodTags (span, path, kind) {
```
It's already tested by the instrumentation for the client and server. Let's add tests for `getFilter` when we extract it since right now it's duplicated in a few places.

rishabh replied:
I was thinking we could have a few test cases since right now we only test for when the path is `/test.TestService/<method>`, but I think it can get a bit more complicated than that, though.

We also definitely need tests for `getFilter` 👍

rochdev commented on packages/datadog-plugin-grpc/src/client.js:
```diff
@@ -0,0 +1,177 @@
+
+          proto[method] = wrapMethod(tracer, config, proto[method], methods[method], grpc)
+
+          if (originalName) {
```
Fixed with a check in `wrapMethod`.

rochdev commented on packages/datadog-plugin-grpc/src/client.js:
```diff
@@ -0,0 +1,177 @@
+          proto[method] = wrapMethod(tracer, config, proto[method], methods[method], grpc)
+
+          if (originalName) {
+            proto[originalName] = wrapMethod(tracer, config, proto[originalName], methods[originalName], grpc)
```
Fixed with a check in `wrapMethod`.

rishabh commented on packages/datadog-plugin-grpc/src/client.js:
```diff
@@ -0,0 +1,181 @@
+  const parent = scope.active()
+
+  return function (err) {
+    err && span.setTag('error', err)
```
Maybe we should use `Tags.ERROR` instead of the literal value.

rishabh commented on packages/datadog-plugin-grpc/src/client.js:
```diff
@@ -0,0 +1,181 @@
+          proto[method] = wrapMethod(tracer, config, proto[method], methods[method], grpc)
+
+          if (originalName) {
+            proto[originalName] = wrapMethod(tracer, config, proto[originalName], methods[method], grpc)
```
Is there a way to refactor this so we can just assign the same `methodWithTrace` that we assigned to `proto[method]` to `proto[originalName]`. The reason is `gRPC` [does something similar](https://github.com/grpc/grpc-node/blob/2ab15fef4ff19d02b78f33a0290ca10f6978262b/packages/grpc-js/src/make-client.ts#L140).

rishabh replied:
I also tested this with a weird case: if you have both a camelCased and snake_cased method defined, the app just doesn't work properly, so the previous should be okay to do I think.

rishabh commented on packages/datadog-plugin-grpc/src/server.js:
```diff
@@ -0,0 +1,144 @@
+  call.on('error', err => {
+    span.setTag('grpc.status.code', err.code)
+
+    handleError(span, err)
```
We can probably remove `handleError` because it  makes the style look a bit weird since it's just a simple, one-liner function. I'm open to leaving it in.

rishabh commented on packages/datadog-plugin-grpc/src/server.js:
```diff
@@ -0,0 +1,144 @@
+      scope.bind(call)
+
+      // Finish the span if the call was cancelled.
+      call.on('cancelled', () => {
```
If we're removing event listeners for when a span finishes, we'll need a reference to this listener.

Also I think we can add these listeners on a `.once()`.

rishabh commented on packages/datadog-plugin-grpc/src/server.js:
```diff
@@ -0,0 +1,144 @@
+    span.finish()
+
+    if (callback) {
+      scope.bind(callback, childOf).apply(this, arguments)
```
I think we should return this.

rishabh commented on packages/datadog-plugin-grpc/src/server.js:
```diff
@@ -0,0 +1,144 @@
+  config = config.server || config
+
+  return function wrapRegister (register) {
+    return function registerWithTrace (name, handler, serialize, deserialize, type) {
```
We don't use the last 3 parameters so we can just get rid of em.

rishabh commented on packages/datadog-plugin-grpc/src/server.js:
```diff
@@ -0,0 +1,144 @@
+        span.setTag('grpc.status.code', err.code)
+      }
+
+      handleError(span, err)
```
If we're removing `handleError` because of the style (commented previously), then we'll need to change this, too.

rishabh commented on packages/datadog-plugin-grpc/src/server.js:
```diff
@@ -0,0 +1,144 @@
+const { addMethodTags, addMetadataTags, getFilter } = require('./util')
+
+function handleError (span, err) {
+  span.setTag('error', err)
```
I think we can just do a `span.setTag(Tags.ERROR, err)` in this case.

rishabh commented on packages/datadog-plugin-grpc/src/server.js:
```diff
@@ -0,0 +1,144 @@
+
+      // Finish the span if the call was cancelled.
+      call.on('cancelled', () => {
+        span.setTag('grpc.status.code', grpc.status.CANCELLED)
```
A small style change, but do you want to make `grpc.status.code` into a constant.

ebenoist commented:
👍
